sphinx/builders/linkcheck.py

```diff
@@ -268,53 +283,99 @@ def write_linkstat(self, data: dict) -> None:
         self.json_outfile.write('\n')
 
     def finish(self) -> None:
+        checker = HyperlinkAvailabilityChecker(self.env, self.config, self)
         logger.info('')
 
         with open(path.join(self.outdir, 'output.txt'), 'w') as self.txt_outfile,\
              open(path.join(self.outdir, 'output.json'), 'w') as self.json_outfile:
-            total_links = 0
-            for hyperlink in self.hyperlinks.values():
-                if self.is_ignored_uri(hyperlink.uri):
-                    self.process_result(
-                        CheckResult(hyperlink.uri, hyperlink.docname, hyperlink.lineno,
-                                    'ignored', '', 0))
-                else:
-                    self.wqueue.put(hyperlink, False)
-                    total_links += 1
-
-            done = 0
-            while done < total_links:
-                self.process_result(self.rqueue.get())
-                done += 1
+            for result in checker.check(self.hyperlinks):
+                self.process_result(result)
 
         if self._broken:
             self.app.statuscode = 1
 
+
+class HyperlinkAvailabilityChecker:
+    def __init__(self, env: BuildEnvironment, config: Config,
+                 builder: CheckExternalLinksBuilder = None) -> None:
+        # Warning: builder argument will be removed in the sphinx-5.0.
+        # Don't use it from extensions.
+        # tag: RemovedInSphinx50Warning
```

Comment:
This could raise a `RemovedInSphinx50Warning` when builder is not None. (same remark for `HyperlinkAvailabilityCheckerWorker`)

Reply:
At present, our `CheckExternalLinksBuilder` passes the builder argument always. It means linkcheck builder will raise a `RemovedInSphinx50Warning`.  I think it's very noisy for users.

Reply:
Make sense, thanks for explaining.